text/0000-interrupt-calling-conventions.md

```diff
@@ -89,61 +89,67 @@ By using these ABIs, it is possible to implement interrupt handlers directly in
 # Reference-level explanation
 [reference-level-explanation]: #reference-level-explanation
 
-- lowered to llvm
-- functions should not be called by software
-- functions are safe, but not fool-proof
-- some calling conventions are only available for certain platforms
-  - e.g.: `x86-interrupt`: no red-zone, no SSE?
+The exact requirements and properties of the different interrupt calling conventions must be defined and documented before stabilizing them. However, there are some properties and requirements that apply to all interrupt calling conventions.
 
-<!--
-This is the technical portion of the RFC. Explain the design in sufficient detail that:
+## Compiler Checks
+Interrupt calling conventions have strict requirements that are checked by the Rust compiler:
 
-- Its interaction with other features is clear.
-- It is reasonably clear how the feature would be implemented.
-- Corner cases are dissected by example.
+- They must not be called by Rust code.
+- The function signature must satisfy the requirements.
+- They are only available on specific targets and might require specific target settings.
+- All other requirements imposed by the implementation of the calling convention in LLVM.
 
-The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
--->
+If any of these conditions are violated, the compiler throws an error. It should not be possible to cause LLVM errors using interrupt calling conventions.
+
+## Stability
+Since interrupt calling conventions are closely tied to a target architecture, they are only as stable as the corresponding target triple, even if the interrupt calling convention is stabilized. If support for a target triple is removed from Rust, removing support for corresponding interrupt calling conventions is _not_ considered a breaking change.
+
+Apart from this limitation, interrupt calling conventions fall under Rust's normal stability guarantees. For this reason, special care must be taken before stabilizing interrupt calling conventions that are implemented outside of `rustc` (e.g. in LLVM).
+
+## Safety
```

Comment:
This needs to say that interrupt routines are implicitly unsafe. Even if the interrupt routine only calls safe Rust functions, it can easily observe intermediate CPU states that are beyond the Rust safety model, or corrupt the state so that undefined behavior occurs after the interrupt routine returns.

Reply:
Could you clarify what you mean by "implictly unsafe"? That they should be treated like an `unsafe fn` so that no `unsafe {}` blocks are required in it?

I agree this question is something that we should specify in the RFC in more detail. My opinion on this is that we should keep the function body safe, but make all dangerous operations unsafe:

- Functions with interrupt calling conventions cannot be called from Rust code directly, so there is no way to cause undefined behavior using only safe code.
- To register a interrupt handler function on the CPU, you typically need some unsafe code (e.g. inline asm) or place it at a specific address using a linker script (which allow all kinds of unsafe operations). In this "unsafe" code, you need to make sure that the referenced function fulfills all necessary safety invariants.
- We should try to make all operations that corrupt the CPU state unsafe, if possible. For example, the `x86_64` crate requires an `unsafe` block to [modify the passed `InterruptStackFrame`](https://docs.rs/x86_64/0.14.8/x86_64/structures/idt/struct.InterruptStackFrame.html#method.as_mut). Ideally, this `unsafe` block should be required directly by the compiler, for example by requiring raw pointer arguments instead of references for the `x86-interrupt` calling convention.

I'm happy to discuss this further!

Reply:
What I meant is that even if the function is not marked as unsafe, the usual rules about Rust memory safety do not apply at all to these functions. There are a lot of restrictions that have to be followed, depending on the rest of the system, and I'm not sure if Rust is actually up for that. (With C or C++, separate compilation with different compiler flags is sometimes used.)

Reply:
That seems like a case by case basis as to how safe each ABI is going to be. The interrupt handler function ABI *could* (for example) have additional pre and post code inserted by rustc automatically so that the user's source code always runs in a safe environment.

Like we can literally make up these ABIs to do whatever, since normal code is never going to interface with the interrupt handler code.

Reply:
@fweimer can you elaborate a bit more about what rules re: Rust memory safety are inapplicable? The function doesn't get access to anything but static data, right?

Reply:
The function in fact *could* get access to the register values the main program had when the interrupt happened. The function also *might* get far far less stack space than normal, and would likely not have guard pages set up for this stack. There's a lot of potential interrupt styles.

Reply:
@Lokathor how could it get access to those things, specifically?

Reply:
This is quite "into the weeds", but for example on ARM 32-bit there's multiple CPU modes, and an interrupt will make the CPU switch modes and jump to the interrupt handler, but separate modes *mostly* share all the same registers. So your interrupt handler just naturally sees the same registers as the main program happened to have at the time the interrupt happened. In rust terms this doesn't involve unsafe at all. In fact, for a software interrupt this is intentional. However, this does mean that the handler has to properly preserve all register values and restore them before ending the interrupt.

Reply:
Ah, I see your point. Well then at least the situations I was talking about are fine enough if the compiler is handling it. Though, perhaps **fweimer** had something else in mind.

Reply:
Once you introduce interrupts you are effectively no longer a single-threaded process as the easiest way to model interrupts in the rust abstract machine is as running on a separate thread which could synchronize with the non-interrupt thread using regular concurrency methods. It just so happens that while the interrupt runs, the non-interrupt thread doesn't make any progress and thus you may deadlock if you wait for it.

Reply:
I don't think it can be treated the same as a separate thread precisely because it blocks the non-interrupt code while running.

I've equated it to a signal handler before, which is far more limited in what it can do.